src/robin_dict.jl

```diff
@@ -0,0 +1,682 @@
+        h.vals[index] = cval
+        return index
+    end
+
+    @inbounds if h.slots[index] == 0x0
+    	h.count += 1
+    end
+
+    @inbounds h.slots[index] = 0x1
+    @inbounds h.vals[index] = cval
+    @inbounds h.keys[index] = ckey
+    
+    @assert cdibs > 0
+    @inbounds h.dibs[index] = cdibs
+    
+    h.totalcost += 1
+    h.maxprobe = max(h.maxprobe, cdibs)
+    if h.idxfloor == 0
+        h.idxfloor = index
+    else
+        h.idxfloor = min(h.idxfloor, index)
+    end
+    return index
+end
+
+#same as rh_insert but totalcost doesn't increase here
+function rh_insert_for_rehash!(h::RobinDict{K, V}, key::K, val::V) where {K, V}
+    # table full
+    if h.count == length(h.keys)
+        return -1
+    end
+    ckey, cval, cdibs = key, val, 1
+    sz = length(h.keys)
+    index = hashindex(ckey, sz)
+    @inbounds while true
+        if (h.slots[index] == 0x0) || (h.slots[index] == 0x1 && h.keys[index] == ckey)
+            break
+        end
+        if h.dibs[index] < cdibs
+            h.vals[index], cval = cval, h.vals[index]
+            h.keys[index], ckey = ckey, h.keys[index]
+            h.maxprobe = max(h.maxprobe, cdibs)
+            h.dibs[index], cdibs = cdibs, h.dibs[index]
+        end
+        cdibs += 1
+        index = (index & (sz - 1)) + 1
+    end
+    
+    @inbounds if h.slots[index] == 0x1 && h.keys[index] == ckey
+        h.vals[index] = cval
+        return index
+    end
+
+    @inbounds if h.slots[index] == 0x0
+        h.count += 1
+    end
+
+    @inbounds h.slots[index] = 0x1
+    @inbounds h.vals[index] = cval
+    @inbounds h.keys[index] = ckey
+    
+    @assert cdibs > 0
+    @inbounds h.dibs[index] = cdibs
+    
+    h.maxprobe = max(h.maxprobe, cdibs)
+    if h.idxfloor == 0
+        h.idxfloor = index
+    else
+        h.idxfloor = min(h.idxfloor, index)
+    end
+    return index
+end
+
+#rehash! algorithm
+function rehash!(h::RobinDict{K,V}, newsz = length(h.keys)) where {K, V}
+    olds = h.slots
+    oldk = h.keys
+    oldv = h.vals
+    oldd = h.dibs
+    sz = length(olds)
+    newsz = _tablesz(newsz)
+    h.totalcost += 1
+    h.idxfloor = 1
+    if h.count == 0
+        resize!(h.slots, newsz)
+        fill!(h.slots, 0)
+        resize!(h.keys, sz)
+        resize!(h.vals, sz)
+        resize!(h.dibs, sz)
+        h.count = 0
+        h.maxprobe = 0
+        h.totalcost = 0
+        h.idxfloor = 0
+        return h
+    end
+
+    h.slots = zeros(UInt8,newsz)
+    h.keys = Vector{K}(undef, newsz)
+    h.vals = Vector{V}(undef, newsz)
+    h.dibs = Vector{Int}(undef, newsz)
+    fill!(h.dibs, 0)
+    totalcost0 = h.totalcost
+    h.count = 0
+    h.maxprobe = 0
+    h.idxfloor = 0
+
+    for i = 1:sz
+        @inbounds if olds[i] == 0x1
+            k = oldk[i]
+            v = oldv[i]
+            rh_insert_for_rehash!(h, k, v)
+            if h.totalcost != totalcost0
+                # if `h` is changed by a finalizer, retry
+                return rehash!(h, newsz)
```

Comment:
you've already destroyed `h` by assigning the fields at the top of the function, you need to assign those at the end